### PR TITLE
enable readiness reflector for Standalone NEG

### PR DESCRIPTION
**Pull request by freehan**
Enable Pod Readiness Gate for Standalone NEG

### CODE REVIEW

File: pkg/neg/readiness/poller.go [hunk 1]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
-		for _, r := range res {
-			healthy, err := p.processHealthStatus(key, r)
-			if healthy && err == nil {
-				healthyCount++
-			}
+	return p.processHealthStatus(key, res)
+}
+
+// processHealthStatus processes the health status response from NEG API.
+// It processed health status for each endpoint and signal patcher correspondingly in the following conditions:
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]:
```
// processHealthStatus updates Pod readiness gates in response to a health status update.
//
// We update the pod (using the patcher) when:
// 1. ...
```

File: pkg/neg/readiness/poller.go [hunk 2]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+// 1. if the endpoint considered healthy with one of the backend service health check
+// 2. if the NEG is not associated with any health checks
+// It returns true if retry is needed.
+func (p *poller) processHealthStatus(key negMeta, healthStatuses []*composite.NetworkEndpointWithHealthStatus) (bool, error) {
+	var retry bool
+	errList := []error{}
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: `var errList []error`

File: pkg/neg/readiness/poller.go [hunk 3]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+	return p.processHealthStatus(key, res)
+}
+
+// processHealthStatus processes the health status response from NEG API.
+// It processed health status for each endpoint and signal patcher correspondingly in the following conditions:
+// 1. if the endpoint considered healthy with one of the backend service health check
+// 2. if the NEG is not associated with any health checks
+// It returns true if retry is needed.
+func (p *poller) processHealthStatus(key negMeta, healthStatuses []*composite.NetworkEndpointWithHealthStatus) (bool, error) {
+	var retry bool
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: these var decls are kind of random, prefer putting locking at the top and grouping everything

```
p.lock.Lock()
defer p.lock.Unlock()
var (
  retry bool
  errList []error
  patchCount int
  unhealthyPods []types.NamespacedName
  // healthChecked indicates ...
  healthChecked bool
)
```

Comment (bowei) [inline pkg/neg/readiness/poller.go]: `retry` variable does not look like it's actually needed

File: pkg/neg/readiness/poller.go [hunk 4]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+
+		podName, ok := p.getPod(key, ne)
+		if !ok {
+			// The pod is not in interest. Skip
+			continue
+		}
+
+		bsKey := getHealthyBackendService(healthStatus)
+		if bsKey == nil {
+			unhealthyPods = append(unhealthyPods, podName)
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: if bsKey == nil { 
  unhealthyPods = append(...)
 continue 
}

File: pkg/neg/readiness/poller.go [hunk 5]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+	// health check to start. The assumption is that if at least one of the endpoint in a NEG has
+	// health status, all the endpoints in the NEG should have health status eventually.
+	var healthChecked bool
+	// patchCount is the count of the pod got patched
+	var patchCount int
+	unhealthyPods := []types.NamespacedName{}
+
+	for _, healthStatus := range healthStatuses {
+		if healthStatus == nil {
+			continue
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: why would the healthStatus be nil? Also, you should log a warn or error

File: pkg/neg/readiness/poller.go [hunk 6]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+	// healthChecked indicates whether at least one of the endpoint in response has health status.
+	// If a NEG is attached to a Backend Service with health check, all endpoints
+	// in the NEG will be health checked. However, for new endpoints, it may take a while for the
+	// health check to start. The assumption is that if at least one of the endpoint in a NEG has
+	// health status, all the endpoints in the NEG should have health status eventually.
+	var healthChecked bool
+	// patchCount is the count of the pod got patched
+	var patchCount int
+	unhealthyPods := []types.NamespacedName{}
+
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: Is there any harm in adding lots of logging at V(3) or above?
```
glog.V(3).Infof("processHealthStatus(%v, %+v)", ...)
etc...
glog.V(3).Infof("processing healthStatus %v ", ...)
```

File: pkg/neg/readiness/poller.go [hunk 7]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
 				errList = append(errList, err)
+				continue
+			} else {
+				patchCount++
 			}
 		}
-		if healthyCount != len(p.pollMap[key].endpointMap) {
-			retry = true
+	}
+	// if the NEG is not health checked, signal the patcher to mark the unhealthy pods to be Ready.
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: Why does this happen? I think this needs an explanation.
(NEG not healthchecked => unhealthy pod = ready)

File: pkg/neg/readiness/poller.go [hunk 8]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
-	ne := negtypes.NetworkEndpoint{
-		IP:   healthStatus.NetworkEndpoint.IpAddress,
-		Port: strconv.FormatInt(healthStatus.NetworkEndpoint.Port, 10),
-		Node: healthStatus.NetworkEndpoint.Instance,
 	}
-	podName, ok := p.getPod(key, ne)
-	if !ok {
-		return false, nil
+
+	// if the number of patch operation is more than the endpoint in question.
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: the comment seems to say patchCount > len(...) but the code says the opposite.
I would change the comment to say "// If we didn't patch all of the endpoints, we must keep polling for health status " or something equivalent.

File: pkg/neg/readiness/poller.go [hunk 9]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+	//
+	if patchCount < len(p.pollMap[key].endpointMap) {
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]:
```
needsRetry := patchCount < len(p.pollMap[key].endpointMap)
return needsRetry, utilerrors.NewAggregate(errList)
```

File: pkg/neg/readiness/poller.go [hunk 10]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
 
+// getHealthyBackendService returns one of the backend service key where the endpoint is considered healthy.
+func getHealthyBackendService(healthStatus *composite.NetworkEndpointWithHealthStatus) *meta.Key {
 	for _, hs := range healthStatus.Healths {
 		if hs == nil {
 			continue
 		}
 		if hs.BackendService == nil {
-			klog.Warningf("Backend service is nil in health status of network endpoint %v: %v", ne, hs)
+			klog.Warningf("Backend service is nil in health status of network endpoint %v", healthStatus)
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: this should be an Error

File: pkg/neg/readiness/poller.go [hunk 11]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+	if patchCount < len(p.pollMap[key].endpointMap) {
+		retry = true
 	}
+	return retry, utilerrors.NewAggregate(errList)
+}
 
+// getHealthyBackendService returns one of the backend service key where the endpoint is considered healthy.
+func getHealthyBackendService(healthStatus *composite.NetworkEndpointWithHealthStatus) *meta.Key {
 	for _, hs := range healthStatus.Healths {
 		if hs == nil {
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: this should be an Error log.
Also, kind of strange to have to check for nils everywhere

File: pkg/neg/readiness/poller.go [hunk 12]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
-		// This assumes the ingress backend service uses the NEG naming scheme. Hence the backend service share the same name as NEG.
-		if strings.Contains(hs.BackendService.BackendService, key.Name) {
-			if hs.HealthState == healthyState {
-				healthy = true
-				err := p.patcher.syncPod(keyFunc(podName.Namespace, podName.Name), key.Name)
-				return healthy, err
+		if hs.HealthState == healthyState {
+			id, err := cloud.ParseResourceURL(hs.BackendService.BackendService)
+			if err != nil {
+				klog.Errorf("Failed to parse backend service reference from a Network Endpoint health status %v: %v", healthStatus, err)
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: continue?

File: pkg/neg/readiness/poller.go [hunk 13]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+
+	// if the number of patch operation is more than the endpoint in question.
+	//
+	if patchCount < len(p.pollMap[key].endpointMap) {
+		retry = true
 	}
+	return retry, utilerrors.NewAggregate(errList)
+}
 
+// getHealthyBackendService returns one of the backend service key where the endpoint is considered healthy.
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: // ... returns the first healthy backend service key

File: pkg/neg/readiness/poller.go [hunk 14]
```diff
@@ -135,69 +137,125 @@ func (p *poller) Poll(key negMeta) (retry bool, err error) {
...
+				return id.Key
 			}
 		}
+	}
+	return nil
+}
 
+// hasHealthStatus returns true if there is at least 1 health status associated with the endpoint.
+func hasHealthStatus(healthStatus *composite.NetworkEndpointWithHealthStatus) bool {
+	if healthStatus != nil && len(healthStatus.Healths) > 0 {
```
Comment (bowei) [inline pkg/neg/readiness/poller.go]: return healthStatus != nil && len(...)

File: pkg/neg/readiness/poller.go (hunk 5 above)
Comment (freehan) [reply pkg/neg/readiness/poller.go]: Just a sanity check since the Go Client parsing GCE API response is not bullet proof. 
Will log a warning.

**Comment by freehan** [comment]
Fixed the comments

**Pull request merged**